op-node/p2p/peer_scorer.go

```diff
@@ -1,16 +1,92 @@
 package p2p
 
 import (
+	"fmt"
+	"strconv"
+	"strings"
+
 	log "github.com/ethereum/go-ethereum/log"
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	peer "github.com/libp2p/go-libp2p/core/peer"
 )
 
 type scorer struct {
-	peerStore Peerstore
-	metricer  GossipMetricer
-	log       log.Logger
-	gater     PeerGater
+	peerStore           Peerstore
+	metricer            GossipMetricer
+	log                 log.Logger
+	gater               PeerGater
+	bandScoreThresholds BandScorer
+}
+
+// bandScoreThresholds holds the thresholds for classifying peers
+// into different score bands.
+type bandScoreThresholds struct {
+	bands      map[string]float64
+	lowestBand string
+}
+
+// BandScorer is an interface for placing peer scores
+// into various bands.
+//
+// Implementations are expected to construct internals using the
+// [Parse] function and then expose the [Bucket] function for
+// downstream [BandScorer] consumers.
+//
+//go:generate mockery --name BandScorer --output mocks/
+type BandScorer interface {
+	Parse(str string) error
+	Bucket(score float64) string
+	Reset()
+}
+
+// NewBandScorer constructs a new [BandScorer] instance.
+func NewBandScorer() BandScorer {
+	return &bandScoreThresholds{
+		bands: make(map[string]float64),
+	}
+}
+
+// Reset wipes the internal state of the [BandScorer].
+func (s *bandScoreThresholds) Reset() {
+	s.bands = make(map[string]float64)
+}
+
+// Parse creates a [BandScorer] from a given string.
+func (s *bandScoreThresholds) Parse(str string) error {
+	var lowestThreshold float64
+	for i, band := range strings.Split(str, ";") {
+		// Skip empty band strings.
+		band := strings.TrimSpace(band)
+		if band == "" {
+			continue
+		}
+		split := strings.Split(band, ":")
+		if len(split) != 2 {
+			return fmt.Errorf("invalid score band: %s", band)
+		}
+		threshold, err := strconv.ParseFloat(split[0], 64)
+		if err != nil {
+			return err
+		}
+		s.bands[split[1]] = threshold
+		if threshold < lowestThreshold || i == 0 {
+			s.lowestBand = split[1]
+			lowestThreshold = threshold
+		}
+	}
+	return nil
+}
+
+// Bucket returns the appropriate band for a given score.
+func (s *bandScoreThresholds) Bucket(score float64) string {
+	for band, threshold := range s.bands {
+		if score >= threshold {
+			return band
+		}
+	}
+	// If there is no band threshold lower than the score,
+	// the peer must be placed in the lowest bucket.
+	return s.lowestBand
```

Comment:
This doesn't work unfortunately because map iteration in Go is in random order. It's a contentious "feature" of Go.

In the constructor/`Parse` you have to record the name-threshold pairs in a slice to get an order and then iterate over that slice so that you return the highest threshold that is >= the score.

Please also add a test for this, it would have failed here.

 The constructor should also validate that the thresholds are ordered. This should also be tested.